mne/export/_edf.py

```diff
@@ -0,0 +1,225 @@
+            ch_types_phys_max[_type] = _data.max()
+            ch_types_phys_min[_type] = _data.min()
+    else:
+        # get the physical min and max of the data in uV
+        # Physical ranges of the data in uV is usually set by the manufacturer
+        # and properties of the electrode. In general, physical max and min
+        # should be the clipping levels of the ADC input and they should be
+        # the same for all channels. For example, Nihon Kohden uses +3200 uV
+        # and -3200 uV for all EEG channels (which are the actual clipping
+        # levels of their input amplifiers & ADC).
+        # For full discussion, see: https://github.com/sccn/eeglab/issues/246
+        pmin, pmax = physical_range[0], physical_range[1]
+
+        # check that physical min and max is not exceeded
+        if data.max() > pmax:
+            raise RuntimeError(f'The maximum μV of the data {data.max()} is '
+                               f'more than the physical max passed in {pmax}.')
+        if data.min() < pmin:
+            raise RuntimeError(f'The minimum μV of the data {data.min()} is '
+                               f'less than the physical min passed in {pmin}.')
+
+    # create instance of EDF Writer
+    hdl = EDFwriter(fname, file_type, n_channels)
+
+    # set channel data
+    for idx, ch in enumerate(ch_names):
+        if physical_range == 'auto':
+            # take the channel type minimum and maximum
+            ch_type = ch_types[idx]
+            pmin, pmax = ch_types_phys_min[ch_type], ch_types_phys_max[ch_type]
+
+        for key, val in [('PhysicalMaximum', pmax),
+                         ('PhysicalMinimum', pmin),
+                         ('DigitalMaximum', digital_max),
+                         ('DigitalMinimum', digital_min),
+                         ('PhysicalDimension', phys_dims),
+                         ('SampleFrequency', out_sfreq),
+                         ('SignalLabel', ch),
+                         ('PreFilter', filter_str_info)]:
+            _try_to_set_value(hdl, key, val, channel_index=idx)
+
+    # set patient info
+    subj_info = raw.info.get('subject_info')
+    if subj_info is not None:
+        birthday = subj_info.get('birthday')
+
+        # get the full name of subject if available
+        first_name = subj_info.get('first_name')
+        last_name = subj_info.get('last_name')
+        first_name = first_name or ''
+        last_name = last_name or ''
+        joiner = ''
+        if len(first_name) and len(last_name):
+            joiner = ' '
+        name = joiner.join([first_name, last_name])
+
+        hand = subj_info.get('hand')
+        sex = subj_info.get('sex')
+
+        if birthday is not None:
+            if hdl.setPatientBirthDate(birthday[0], birthday[1],
+                                       birthday[2]) != 0:
+                raise RuntimeError(f"Setting Patient Birth Date to {birthday} "
+                                   f"returned an error")
```

Comment:
See: https://github.com/mne-tools/mne-python/pull/9643#discussion_r694945767

I don't think this error would ever actually occur, unless someone hacked MNE to make the birthdate non-compliant in the first place. MNE has checks for these types of metadata already in place in Raw.

Reply:
So why are we re-raising the error then if we cannot be more specific? I would not check for this error if it is already checked elsewhere.

Reply:
So the issue is that EDFLib-Python doesn't actually raise an error. It will just return `0` for error and `1` for success.

Reply:
Assuming we meet the specifications of EDF when setting up our data structures, then everything should work. I guess, I meant to say, these error checks are there to make sure that at least the user is aware that there was some buggy implementation change on our end that strayed away from EDF, and thus export EDF stopped working.

Else just returning 0 would not result in a Python error.

Reply:
If you want, I can actually test the errors here:

https://gitlab.com/Teuniz/EDFlib-Python/-/blob/master/src/EDFlib/edfwriter.py#L796

or here

https://gitlab.com/Teuniz/EDFlib-Python/-/blob/master/src/EDFlib/edfwriter.py#L581

For example, by creating a faulty annotation/birthday.

Reply:
Lmk wyt?

Reply:
Yes, it would be interesting to see what happens. Does the EDF file still get created?

In any case, why don't you use `_try_to_set_value()` in these cases? This already raises an error if the return value is not equal to zero.

Reply:
`_try_to_set_value` only works on one argument functions. I wasn't sure how to generalize that function to make it for these 2 cases :/.

I'll add some tests for faulty annotations/birthdates.

Reply:
I can test birth date and measurement date with non-compliant EDF metadata, but Annotations implicitly only has errors that are already checked and controlled for in Raw, so I can't actually even create an error:

```
    # add a bad annotation
    annots = Annotations(onset=-1, duration=0, description='test')
    raw = RawArray(data, info)
    raw.set_annotations(annots)
    with pytest.raises(RuntimeError, match='writeAnnotation()'):
        raw.export(temp_fname)
```

^ won't even work cuz the annotation is dropped.

Reply:
I've updated the test file to reflect these two additional tests